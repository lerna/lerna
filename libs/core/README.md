# core

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test core` to execute the unit tests via [Jest](https://jestjs.io).

## Running lint

Run `nx lint core` to execute the lint via [ESLint](https://eslint.org/).
