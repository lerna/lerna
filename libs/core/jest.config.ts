/* eslint-disable */
export default {
  displayName: "core",
  preset: "../../jest.preset.js",
  globals: {
    "ts-jest": {
      tsconfig: "<rootDir>/tsconfig.spec.json",
    },
  },
  testEnvironment: "node",
  transform: {
    "^.+\\.[tj]sx?$": "ts-jest",
  },
  // Needed to add "json" to avoid issue resolving spdx-license-ids
  moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json"],
  coverageDirectory: "../../coverage/libs/core",
  testTimeout: 45e3,
};
