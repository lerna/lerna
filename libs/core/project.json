{
  "name": "core",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/core/src",
  "projectType": "library",
  "implicitDependencies": ["!test-helpers"],
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/core/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/core/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
