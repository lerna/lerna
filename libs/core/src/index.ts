export { checkWorkingTree, throwIfUncommitted } from "./lib/check-working-tree";
export * from "./lib/cli";
export * from "./lib/collect-updates";
export { Command } from "./lib/command";
export { applyBuildMetadata, recommendVersion, updateChangelog } from "./lib/conventional-commits";
export { createSymlink } from "./lib/create-symlink";
export { describeRef } from "./lib/describe-ref";
export { filterOptions, getFilteredPackages } from "./lib/filter-options";
export { hasNpmVersion } from "./lib/has-npm-version";
export { listableFormat } from "./lib/listable/listable-format";
export { listableOptions } from "./lib/listable/listable-options";
export { logPacked } from "./lib/log-packed";
export { npmInstall, npmInstallDependencies } from "./lib/npm-install";
export { npmPublish } from "./lib/npm-publish";
export { npmRunScript, npmRunScriptStreaming } from "./lib/npm-run-script";
export { getOneTimePassword } from "./lib/otplease";
export { output } from "./lib/output";
export { packDirectory } from "./lib/pack-directory";
export { Package } from "./lib/package";
export { PackageGraph } from "./lib/package-graph";
export { prereleaseIdFromVersion } from "./lib/prerelease-id-from-version";
export { generateProfileOutputPath, Profiler } from "./lib/profiler";
export { getPackages, Project } from "./lib/project";
export { promptConfirmation, promptSelectOne, promptTextInput } from "./lib/prompt";
export { pulseTillDone } from "./lib/pulse-till-done";
export { rimrafDir } from "./lib/rimraf-dir";
export { createRunner, runLifecycle } from "./lib/run-lifecycle";
export { runTopologically } from "./lib/run-topologically";
export { createGitHubClient, createGitLabClient, parseGitRepo } from "./lib/scm-clients";
export { symlinkBinary } from "./lib/symlink-binary";
export { symlinkDependencies } from "./lib/symlink-dependencies";
export { default as tempWrite } from "./lib/temp-write";
export { timer } from "./lib/timer";
export { ValidationError } from "./lib/validation-error";
export { PackageGraphNode } from "./lib/package-graph/package-graph-node";
export { CommandConfigOptions } from "./lib/project";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const npmConf = require("./lib/npm-conf");
// eslint-disable-next-line @typescript-eslint/no-var-requires
const npmDistTag = require("./lib/npm-dist-tag");
export { npmConf, npmDistTag };
