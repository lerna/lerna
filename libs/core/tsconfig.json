{
  "extends": "../../tsconfig.base.json",
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ],
  "compilerOptions": {
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true
  }
}
