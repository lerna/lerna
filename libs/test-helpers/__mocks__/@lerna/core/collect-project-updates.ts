/* eslint-disable @typescript-eslint/no-var-requires */
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

const { collectProjects } = jest.requireActual("@lerna/core");

// collectUpdates.setUpdated(cwd, packageNames...)
// otherwise, enables everything
const updated = new Map();

const mockCollectUpdates = jest.fn((filteredProjects, projectGraph, { cwd }) => {
  const targets = updated.get(cwd);
  const updates = targets
    ? new Map(targets.map((name) => [name, projectGraph.nodes[name]]))
    : new Map(Object.entries(projectGraph.nodes));

  return Array.from(updates.values());
});

const setUpdated = (cwd, ...names) => updated.set(cwd, names);

// isolate tests
afterEach(() => {
  updated.clear();
});

module.exports.collectProjectUpdates = mockCollectUpdates;
module.exports.collectProjectUpdates.setUpdated = setUpdated;
module.exports.collectProjects = collectProjects;
