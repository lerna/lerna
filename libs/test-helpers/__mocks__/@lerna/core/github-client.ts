// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

const releases = new Map();

// keep test data isolated
afterEach(() => {
  releases.clear();
});

const client = {
  repos: {
    createRelease: jest.fn((opts) => {
      releases.set(opts.name, opts);
      return Promise.resolve();
    }),
  },
};

module.exports.createGitHubClient = jest.fn(() => client);
module.exports.createGitHubClient.releases = releases;
module.exports.parseGitRepo = () => ({
  owner: "lerna",
  name: "lerna",
});
