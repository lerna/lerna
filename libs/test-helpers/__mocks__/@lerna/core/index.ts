/* eslint-disable @typescript-eslint/no-var-requires */

const collectUpdates = require("./collect-updates");
const checkWorkingTree = require("./check-working-tree");
const conventionalCommits = require("./conventional-commits");
const { output } = require("./output");
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const prompt = require("./prompt");
const npmDistTag = require("./npm-dist-tag");
const { hasNpmVersion } = require("./has-npm-version");
const { packDirectory } = require("./pack-directory");
const { npmPublish } = require("./npm-publish");
const { runLifecycle, createRunner } = require("./run-lifecycle");
const { createGitLabClient } = require("./gitlab-client");
const { createGitHubClient, parseGitRepo } = require("./github-client");

module.exports = {
  ...jest.requireActual("@lerna/core"),
  output,
  ...collectUpdates,
  ...prompt,
  ...checkWorkingTree,
  ...conventionalCommits,
  createGitLabClient,
  createGitHubClient,
  parseGitRepo,
  npmDistTag,
  hasNpmVersion,
  runLifecycle,
  createRunner,
  npmPublish,
  packDirectory,
  createSymlink: jest.fn(),
  npmRunScript: jest.fn(),
  npmRunScriptStreaming: jest.fn(),
  promptConfirmation: jest.fn(),
  rimrafDir: jest.fn(),
  npmInstall: jest.fn(),
  npmInstallDependencies: jest.fn(),
  getOneTimePassword: jest.fn(),
};
