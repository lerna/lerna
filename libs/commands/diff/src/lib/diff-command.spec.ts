import { getPackages } from "@lerna/core";
import {
  commandRunner,
  gitAdd,
  gitCommit,
  gitInit,
  gitSHASerializer,
  gitTag,
  initFixtureFactory,
} from "@lerna/test-helpers";
import execa from "execa";
import fs from "fs-extra";
import path from "path";

// mocked modules
// eslint-disable-next-line @typescript-eslint/no-var-requires
const childProcess = require("@lerna/child-process");

const initFixture = initFixtureFactory(__dirname);

// file under test
// eslint-disable-next-line @typescript-eslint/no-var-requires
const lernaDiff = commandRunner(require("../command"));

// stabilize commit SHA
expect.addSnapshotSerializer(gitSHASerializer);

describe("DiffCommand", () => {
  // overwrite spawn so we get piped stdout, not inherited
  // TODO: refactor based on TS feedback
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  childProcess.spawn = jest.fn((...args) => execa(...args));

  it("should diff packages from the first commit", async () => {
    const cwd = await initFixture("basic");
    const [pkg1] = await getPackages(cwd);
    const rootReadme = path.join(cwd, "README.md");

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    await pkg1.set("changed", 1).serialize();
    await fs.outputFile(rootReadme, "change outside packages glob");
    await gitAdd(cwd, "-A");
    await gitCommit(cwd, "changed");

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const { stdout } = await lernaDiff(cwd)();
    expect(stdout).toMatchSnapshot();
  });

  it("should diff packages from the most recent tag", async () => {
    const cwd = await initFixture("basic");
    const [pkg1] = await getPackages(cwd);

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    await pkg1.set("changed", 1).serialize();
    await gitAdd(cwd, "-A");
    await gitCommit(cwd, "changed");
    await gitTag(cwd, "v1.0.1");

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    await pkg1.set("sinceLastTag", true).serialize();
    await gitAdd(cwd, "-A");
    await gitCommit(cwd, "changed");

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const { stdout } = await lernaDiff(cwd)();
    expect(stdout).toMatchSnapshot();
  });

  it("should diff a specific package", async () => {
    const cwd = await initFixture("basic");
    const [pkg1, pkg2] = await getPackages(cwd);

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    await pkg1.set("changed", 1).serialize();
    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    await pkg2.set("changed", 1).serialize();
    await gitAdd(cwd, "-A");
    await gitCommit(cwd, "changed");

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const { stdout } = await lernaDiff(cwd)("package-2");
    expect(stdout).toMatchSnapshot();
  });

  it("passes diff exclude globs configured with --ignored-changes", async () => {
    const cwd = await initFixture("basic");
    const [pkg1] = await getPackages(cwd);

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    await pkg1.set("changed", 1).serialize();
    await fs.outputFile(path.join(pkg1.location, "README.md"), "ignored change");
    await gitAdd(cwd, "-A");
    await gitCommit(cwd, "changed");

    // TODO: refactor based on TS feedback
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const { stdout } = await lernaDiff(cwd)("--ignore-changes", "**/README.md");
    expect(stdout).toMatchSnapshot();
  });

  it("should error when attempting to diff a package that doesn't exist", async () => {
    const cwd = await initFixture("basic");
    const command = lernaDiff(cwd)("missing");

    await expect(command).rejects.toThrow("Cannot diff, the package 'missing' does not exist.");
  });

  it("should error when running in a repository without commits", async () => {
    const cwd = await initFixture("basic");

    await fs.remove(path.join(cwd, ".git"));
    await gitInit(cwd);

    const command = lernaDiff(cwd)("package-1");
    await expect(command).rejects.toThrow("Cannot diff, there are no commits in this repository yet.");
  });

  it("should error when git diff exits non-zero", async () => {
    const cwd = await initFixture("basic");

    childProcess.spawn.mockImplementationOnce(() => {
      const nonZero = new Error("An actual non-zero, not git diff pager SIGPIPE") as any;
      nonZero.exitCode = 1;

      throw nonZero;
    });

    const command = lernaDiff(cwd)("package-1");
    await expect(command).rejects.toThrow("An actual non-zero, not git diff pager SIGPIPE");
  });
});
