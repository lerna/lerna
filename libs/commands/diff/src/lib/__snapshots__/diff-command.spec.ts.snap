// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffCommand passes diff exclude globs configured with --ignored-changes 1`] = `
diff --git a/packages/package-1/package.json b/packages/package-1/package.json
index SHA..SHA 100644
--- a/packages/package-1/package.json
+++ b/packages/package-1/package.json
@@ -1,5 +1,5 @@
 {
   "name": "package-1",
-  "changed": 0,
+  "changed": 1,
   "version": "1.0.0"
 }
`;

exports[`DiffCommand should diff a specific package 1`] = `
diff --git a/packages/package-2/package.json b/packages/package-2/package.json
index SHA..SHA 100644
--- a/packages/package-2/package.json
+++ b/packages/package-2/package.json
@@ -1,6 +1,6 @@
 {
   "name": "package-2",
-  "changed": 0,
+  "changed": 1,
   "version": "1.0.0",
   "dependencies": {
     "package-1": "^1.0.0"
`;

exports[`DiffCommand should diff packages from the first commit 1`] = `
diff --git a/packages/package-1/package.json b/packages/package-1/package.json
index SHA..SHA 100644
--- a/packages/package-1/package.json
+++ b/packages/package-1/package.json
@@ -1,5 +1,5 @@
 {
   "name": "package-1",
-  "changed": 0,
+  "changed": 1,
   "version": "1.0.0"
 }
`;

exports[`DiffCommand should diff packages from the most recent tag 1`] = `
diff --git a/packages/package-1/package.json b/packages/package-1/package.json
index SHA..SHA 100644
--- a/packages/package-1/package.json
+++ b/packages/package-1/package.json
@@ -1,5 +1,6 @@
 {
   "name": "package-1",
   "changed": 1,
-  "version": "1.0.0"
+  "version": "1.0.0",
+  "sinceLastTag": true
 }
`;
